### DISCUSSION TITLE
vwc-toggle-button

### DISCUSSION THREAD

**Discussion by YonatanKra** [Vonage/vivid#732] [New Developments]
Hi,
We're starting to think about the `vwc-toggle-button` component.
Somewhat like this:
![image](https://user-images.githubusercontent.com/6459899/113009663-d1679500-9180-11eb-8d36-38c72bc35d89.png)
Taken from here: https://material-ui.com/components/toggle-button/

# Suggested API

## Usage Template
```
<vwc-toggle-buttons-group activeindex="2">
 <vwc-button value="whatever"></vwc-button>
 <vwc-button value="whatever1"></vwc-button>
 <vwc-button value="whatever2"></vwc-button>
 <vwc-button value="whatever3"></vwc-button>
</vwc-toggle-button>
```

### Properties/Attributes
| Name | Type | Description |
| - | - | -|
| `multiple` | `reflected` | Allows for multiple selections |
| `value` | `property` | The value/values of the selected button/s. Will be bound to the buttons (clearing this will clear selected buttons in the view as well) |
| `activeIndex` | `property` | The indices list (array) of the currently toggled buttons |

### Events
Name    | event.detail     | Description
-- | -- | --
`toggle`     | {state: []{ activeIndex, value, active }, toggled: {activeIndex, active} } | An object with the current state of the group as well as the toggled button index and state (active/inactive => true/false)

### Design API
There will be a class set on the selected button/s that will change its look according to what will be decided in Figma.

**Comment by tveinfeld** [comment]
what would you say is the value offered here beyond what can be achieved using a group of vwc-buttons?

**Reply by yinonov** [reply]
this is part of the [volta parity](http://volta-core.vonage.com/storybook/?path=/story/components-buttons--grouped) and seems like highly requested component
![image](https://user-images.githubusercontent.com/15140652/113028253-bb170480-9193-11eb-9e86-51689d8d489d.png)

**Reply by tveinfeld** [reply]
got it.
should we deliver logic or simply design here?

**Reply by yinonov** [reply]
care to explain? I think the purpose of this discussion is to find out

**Reply by tveinfeld** [reply]
I meant, perhaps we can just provide the design for this group of buttons without managing state/mode (multiple and value)

**Reply by yinonov** [reply]
I feel it should eventually output value like in https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_select_multiple

**Comment by yinonov** [comment]
> ## Usage Template
> ```
> <vwc-toggle-buttons>
>  <vwc-button value="whatever"></vwc-button>
>  <vwc-button value="whatever1"></vwc-button>
>  <vwc-button value="whatever2"></vwc-button>
>  <vwc-button value="whatever3"></vwc-button>
> </vwc-toggle-button>
> ```


I would suggest keeping it synced with other components like tab-bar and a dedicated option element to not pollute the original button with more style variants or scoping
```js
<vwc-toggle-button-bar>
  <vwc-toggle-button value="whatever"></vwc-toggle-button>
  <vwc-toggle-button value="whatever1"></vwc-toggle-button>
  <vwc-toggle-button value="whatever2"></vwc-toggle-button>
  <vwc-toggle-button value="whatever3"></vwc-toggle-button>
</vwc-toggle-butto-bar>
> ```

**Reply by YonatanKra** [reply]
@yinonov @tveinfeld @gullerya @JoelGraham93 
Plz cast your vote for approach no. 1 ([as in here](https://github.com/Vonage/vivid/discussions/732#discussion-3299029)) or no. 2 ([as in here](https://github.com/Vonage/vivid/discussions/732#discussioncomment-549610)).

**Reply by YonatanKra** [reply]
1

**Reply by yinonov** [reply]
common ppl upvote the latter! it's fantastish! 🦾

**Reply by gullerya** [reply]
2 `vwc-toggle-button-bar`

**Reply by tveinfeld** [reply]
2 `vwc-toggle-button-bar`.. If we're about to hold state, let's do it using state-holding buttons :)

**Reply by lagjamhero** [reply]
2 - although if this is for volta button-group parity, we also supported text buttons

**Reply by jshenkman** [reply]
Yes, we need to support text as well. VCC love this component

**Comment by YonatanKra** [comment]
Ok, the vote was case for:
```
<vwc-toggle-button-bar>
  <vwc-toggle-button value="whatever"></vwc-toggle-button>
  <vwc-toggle-button value="whatever1"></vwc-toggle-button>
  <vwc-toggle-button value="whatever2"></vwc-toggle-button>
  <vwc-toggle-button value="whatever3"></vwc-toggle-button>
</vwc-toggle-butto-bar>
```
Question about it: instead of building one thing, we need to build two new things (toggle-button-bar and toggle-button)?

I know some ppl think inheritance is an innocent and harmless tool, but I'm not one of them - so please let me know you are convinced we'd like to unnecessarily create another inheritance level.

**Reply by tveinfeld** [reply]
There's already a `<vwc-icon-button-toggle>` component, perhaps we can use it?
I guess we're only dealing with icons, right?

**Reply by gullerya** [reply]
I don't think that inheritance is something you need here.
But indeed there is already a toggle button like Tal wrote.

And moreover, I'd suggest the following:
- `bar` component should not know nor assume nothing about it's __first level (slotted?) children__, but 2 things:
  - that they fire `change` event in case of their 'boolean-state-like' value changed / toggled
  - that they have `value` or something like this, that reflects the state

Now it becomes an open API interop: basically, if that would fit, one may have put even checkboxes into the `bar` and it would successfully manage the state toggling between them (+ reflect the chosen value + event to notify outside consumers).

**Reply by gullerya** [reply]
Last 2 cents for now: as the API candidate it is probably worthy to take the `vwc-icon-button-toggle` API:
- `on` property/attribute [boolean] to reflect the state
- `change` event when `on` is being changed

You'd then only need to require a `value` property/attribute so that the `bar` would be able to say which button is currently `on`. `vwc-icon-button-toggle` do not have `value`... so, well, need to think if we'd like to keep it fully open (and just say to the consumer "drop value attribute or property at runtime, we'll pick it up from there (+ define precedence)" or add value to our own `vwc-icon-button-toggle`.

Phase 2 (P1) you could do it fully customizable, allowing this:
```html
<vwc-toggle-button-bar toggle-source="checked" value-source="someCustomProperty">
...
</vwc-toggle-button-bar>
```

**Reply by yinonov** [reply]
I'd drop my thoughts here on, as always, the matter of convention first and foremost. I mean, if other similar components are already defined in a certain sensible way, we better stick to that. I would even say let's align to the API naming of tabs.
furthermore I believe extending subsystem components is a good thing.

**Reply by yinonov** [reply]
not sure if that's rocking the boat here, but the `select` tag might share similar behavior to what we expect where _multiple_ attribute is set https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select

**Reply by yinonov** [reply]
lastly, I think u'll find `vwc-icon-button-toggle` somewhat limiting. maybe only copy certain needed parts of it?

**Reply by yinonov** [reply]
> There's already a `<vwc-icon-button-toggle>` component, perhaps we can use it?
> I guess we're only dealing with icons, right?

requests from devs show more than only icons but text as well. @jshenkman?

**Comment by yinonov** [comment]
https://user-images.githubusercontent.com/15140652/113768796-e916be80-9728-11eb-8b6a-9c9729c64bc6.mov

**Comment by yinonov** [comment]
follow progress at https://tmp.dev.vivid.vonage.com/753/index.html?path=/story/components-atoms-toggle-buttons-group--basic

**Comment by yinonov** [comment]
consider vertical layout to enable navbar usage.
another reason not to enforce wai-aria attributes on assigned slotted elements - consumer would might want to set navbar related values
![image](https://user-images.githubusercontent.com/15140652/114298408-efa58d00-9abe-11eb-9b41-9c697b94d48f.png)
